File: Python/statistics/data_distribution_and_sampling/bootstrap.py
import pandas as pd
from sklearn.utils import resample
import matplotlib.pyplot as plt

if __name__ == "__main__":
    file = "../../data/statistics/loans_income.csv"
    income = pd.read_csv(file)

    results = []
    for n_repeat in range(1000):
        sample = resample(income)
        results.append(sample.median())
    results_S = pd.Series(results)

    print("Bootstrap statistics:")
    print(f"original median: {income.median()}")
    print(f"bias: {results_S.mean() - income.median()}")
    print(f"std error: {results_S.std()}")

    # density plot original
    ax = income["x"].plot.hist(density=True, figsize=(10, 10))
    income["x"].plot.density(ax=ax)
    plt.show()

    # density plot bootstrapped (looks little weird)
    results_df = pd.DataFrame(results)

    ax = results_df.plot.hist(density=True, figsize=(10, 10))
    results_df.plot.density(ax=ax)
    plt.show()

File: R/Scripts/statistics/data_distribution_and_sampling/bootstrap.R
library(boot)
data <- read.csv("Data/loans_income.csv")

stat_fun <- function(x, idx) median(x[idx,])
boot_object <- boot(data, R=1000, statistic = stat_fun)

# density plot original
hist(data[["x"]], freq = FALSE)
lines(density(data[["x"]]), lwd=3, col="green") 

# density plot bootstrapped
hist(boot_object[["t"]], freq = FALSE)
lines(density(boot_object[["t"]]), lwd=3, col="green") 